### PR TITLE
Deployment on aws (DEV-100)

### PR DESCRIPTION
**Pull request by head1328**

### CODE REVIEW

File: .github/workflows/build-test-and-publish.yml [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+          key: nodeModules-${{ hashFiles('**/yarn.lock') }}
+          restore-keys: |
+            nodeModules-
+      
+      - name: Use Node.js
+        uses: actions/setup-node@v1
+        with:
+          node-version: '14'
+
+      - name: Install dependencies
```
Comment (scflode) [inline .github/workflows/build-test-and-publish.yml:40]: This is ATM only needed for building the final container image. Currently there are no JS tests or dependencies.

File: .github/workflows/build-test-and-publish.yml [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        if: steps.cache.outputs.cache-hit != 'true'
+        run: |
+          cd ./assets
+          yarn install
+          yarn deploy
+          yarn dump
+
+      - name: Cache local Elixir cache
```
Comment (scflode) [inline .github/workflows/build-test-and-publish.yml:48]:
```suggestion
      - name: Cache local Elixir dependencies
```

File: .github/workflows/build-test-and-publish.yml [hunk 3]
```diff
@@ -0,0 +1,188 @@
+name: Build
+
+on:
+  push:
+    branches:
+      - master
```
Comment (scflode) [inline .github/workflows/build-test-and-publish.yml:6]:
```suggestion
      - main
```

File: .github/workflows/build-test-and-publish.yml [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+      - DEV-100_deployment-on-aws
```
Comment (scflode) [inline .github/workflows/build-test-and-publish.yml:7]: I guess this is for testing everything and will get removed.

File: .github/workflows/build-test-and-publish.yml [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+            ./CHANGELOG.md
+            ./Dockerfile
+            ./rel/run.sh
+            ./version.txt
+  
+  test:
+    runs-on: ubuntu-latest
+    name: Test Elixir project
+
+    needs: [ build ]
```
Comment (scflode) [inline .github/workflows/build-test-and-publish.yml:91]: I don't think this is necessary. The only "shared" part is getting the Elixir dependencies. The rest ist distinct. I would propose to fetch the dependencies here as well and decouple the jobs.

File: .github/workflows/build-test-and-publish.yml [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+      - name: Restore dependencies cache
+        uses: actions/cache@v2
+        with:
+          path: deps
+          key: ${{ runner.os }}-mix-${{ hashFiles('**/mix.lock') }}
+          restore-keys: ${{ runner.os }}-mix-
+
+      - name: Compile application
+        run: mix compile
```
Comment (scflode) [inline .github/workflows/build-test-and-publish.yml:122]:
```suggestion
        run: mix do deps.get, compile
```

File: .github/workflows/deploy.yml [hunk 7]
```diff
@@ -0,0 +1,60 @@
+
+name: Deploy
+
+on:
+  workflow_run:
+    workflows: [ "Build" ]
+    branches:
+      - master
```
Comment (scflode) [inline .github/workflows/deploy.yml:8]:
```suggestion
      - main
```

File: .github/workflows/build-test-and-publish.yml [hunk 8]
```diff
@@ -0,0 +1,188 @@
+name: Build
+
+on:
+  push:
+    branches:
+      - master
+      - DEV-100_deployment-on-aws
+  pull_request:
+    branches:
+      - master
```
Comment (scflode) [inline .github/workflows/build-test-and-publish.yml:10]:
```suggestion
      - main
```

File: .github/workflows/deploy.yml [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+      - DEV-100_deployment-on-aws
```
Comment (scflode) [inline .github/workflows/deploy.yml:9]:
```suggestion
```

File: .github/workflows/build-test-and-publish.yml (hunk 4 above)
Comment (scflode) [inline .github/workflows/build-test-and-publish.yml:7]:
```suggestion
```

File: .github/workflows/deploy.yml [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    types:
+      - completed
```
Comment (scflode) [inline .github/workflows/deploy.yml:11]: What does this mean? Just curious.

File: .github/workflows/deploy.yml [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    - name: Set env
+      run: |
+        echo ::set-env name=AWS_REGION::eu-central-1
+        echo ::set-env name=CLUSTER_NAME::dealog-dev
+        echo ::set-env name=SERVICE_NAME::backoffice
+        echo ::set-env name=CONTAINER_NAME::backoffice
+        echo ::set-env name=FULL_IMAGE_NAME::$(cat ./dockerized-backoffice/full_image_name.txt)
+        echo ::set-env name=IMAGE_TAG::$(cat ./backoffice/version.txt)
+    
+    - name: Set production env
```
Comment (scflode) [inline .github/workflows/deploy.yml:33]: Where are all the application env vars (DB, Kafka etc.) set? Didn't see it.

File: .github/workflows/dev-deployment.yml [hunk 12]
```diff
@@ -1,29 +0,0 @@
-name: Dev deployment
```
Comment (scflode) [inline .github/workflows/dev-deployment.yml:1]: I would like to keep this for now.

File: .github/workflows/build-test-and-publish.yml [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+
+    steps:
+    - name: Fetch artifact
+      uses: actions/download-artifact@v2
+      with:
+        name: backoffice
+
+    - name: Set env
+      run: |
+        echo ::set-env name=VERSION::$(cat ./version.txt)
```
Comment (scflode) [inline .github/workflows/build-test-and-publish.yml:154]: Were does this come from?

File: Dockerfile [hunk 14]
```diff
@@ -1,53 +1,21 @@
-FROM elixir:1.10.4-alpine AS build
+FROM alpine:3.12
```
Comment (scflode) [inline Dockerfile:1]: These changes mean you cannot build the Docker image locally without doing all the steps manually. What is the reasoning with removing the multistage build?

**Review by scflode — CHANGES_REQUESTED** [review]

File: .github/workflows/build-test-and-publish.yml (hunk 13 above)
Comment (head1328) [reply .github/workflows/build-test-and-publish.yml:154]: see the artifacts

File: .github/workflows/deploy.yml (hunk 11 above)
Comment (head1328) [reply .github/workflows/deploy.yml:33]: they are part of the ecs configuration (see task-definition.json)

File: Dockerfile (hunk 14 above)
Comment (head1328) [reply Dockerfile:1]: because we have the pipeline with all the dep caches. when you want a local version, we have to use 2 dockerfiles.

File: .github/workflows/deploy.yml (hunk 10 above)
Comment (head1328) [reply .github/workflows/deploy.yml:11]: if worklows build was completed (success), then ...

File: Dockerfile (hunk 14 above)
Comment (scflode) [reply Dockerfile:1]: Jepp. I saw. But can't we work with Docker caching instead? This complicates IMO the workflow a bit. Not sure...

File: .github/workflows/deploy.yml (hunk 11 above)
Comment (scflode) [reply .github/workflows/deploy.yml:33]: Is this file supposed to be in the repo? I don't see it.

File: .github/workflows/deploy.yml (hunk 10 above)
Comment (scflode) [reply .github/workflows/deploy.yml:11]: 👍

File: .github/workflows/deploy.yml (hunk 11 above)
Comment (head1328) [reply .github/workflows/deploy.yml:33]: see download task definition below

File: Dockerfile (hunk 14 above)
Comment (head1328) [reply Dockerfile:1]: i think a cache, based on single packages (node or elixir) is more powerful than a cache based on docker layers.

File: .github/workflows/deploy.yml (hunk 11 above)
Comment (scflode) [reply .github/workflows/deploy.yml:33]: hm... this is not good. But I guess there is no easy way out.

Comment (scflode) [reply .github/workflows/deploy.yml:33]: Can't we somehow make this work with GH secrets?

Comment (head1328) [reply .github/workflows/deploy.yml:33]: the only thing that will change is the cluster_name. at the moment we have only one cluster (kiss-principle).

**Review by scflode — APPROVED** [review]
LGTM. Thank you!

File: rel/run.sh [hunk 15]
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/env sh
```
Comment (scflode) [inline rel/run.sh:1]: One last question: Why is this changed?

Comment (head1328) [reply rel/run.sh:1]: personal preference: because i know the absolute path of the current environment (base alpine docker image) and don't want to search inside $PATH.
